File: src/easy/sort_array_by_parity.zig
const std = @import("std");

const expectEqualSlices = std.testing.expectEqualSlices;

const Allocator = std.mem.Allocator;

pub fn sortArrayByParity(comptime T: type, allocator: Allocator, nums: []const T) ?[]const T {
    var result = allocator.alloc(T, nums.len) catch return null;
    if (result.len < 2) {
        std.mem.copyBackwards(T, result, nums);
        return result;
    }
    var even: usize = 0;
    var odd: usize = nums.len - 1;
    for (nums) |n| {
        if (@mod(n, 2) == 0) {
            result[even] = n;
            even += 1;
        } else {
            result[odd] = n;
            if (odd > 0) {
                odd -= 1;
            }
        }
    }
    return result;
}

File: src/easy/sort-array-by-parity.ts
export const sortArrayByParity = (nums: number[]): number[] => {
    const result = Array(nums.length);
    let even = 0;
    let odd = nums.length - 1;
    nums.forEach((num) => {
        if (num % 2 === 0) {
            result[even] = num;
            even++;
        } else {
            result[odd] = num;
            odd--;
        }
    });
    return result;
};
